### ISSUE TITLE
Add options to generate non-null attributes on Records, Pojos, and interfaces in KotlinGenerator

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#10212]
jOOQ 3.14 introduced a new `KotlinGenerator` similar to the existing `ScalaGenerator`, which produces kotlin code from a database (#6248).

The out of the box implementation generates all data types as `T?`, which is the only correct type in all cases. In SQL, all types can be nullable (e.g. after a left join), despite there being `NOT NULL` constraints.

However, many users want some additional type information attached to `Field<T>` types, running the risk of false negatives, where NPE are possible. We could offer some configuration flags with various degrees of `T` vs `T?` type support:

- [x] The default behaviour is to produce false positives (i.e. every type is `T?`, which they can be, in SQL. e.g. after left joining a non-nullable type, it becomes nullable again). This is never really wrong, but maybe inconvenient. Implemented in #6248
- [x] An option is to generate false negatives only on records and/or POJOs (i.e. only nullable types are `T?`, the true positives, but non-nullable types, which are generated as `T`, would produce false negatives because they can still be `null` after an outer join, a union, etc.). But it would be a user choice, just like the `!!` operator. With this choice, `TableField<R, T?>` references would still always be nullable, only records and/or POJOs are affected, as they're mostly used with `SELECT *` projections of single tables (no joins)

## Tasks for records / POJOs / interfaces non-null attributes

- [x] Generate non-null attributes
- [x] Update constructors
  - [x] No default constructor should be available? https://github.com/jOOQ/jOOQ/issues/14002
- [x] Handle embeddables
  - [x] https://github.com/jOOQ/jOOQ/issues/14003
- [x] Interaction with DAOs

## Potential additional work, out of scope for now:

- `DSLContext.newRecord(Table<?>)` could throw NPEs if `NULL` values aren't acceptable, effectively requiring to call `DSLContext.newRecord(Table<?>, Object)`, instead.
- There might be other runtime checks that could help fail early

## Options that will not be implemented as of this issue but are left here for future reference:

The options that used to be documented here have been moved to a separate issue: https://github.com/jOOQ/jOOQ/issues/13999

**Comment by KasperKiiskinen** [comment]
We would really like to at least Records to respect the NOT NULL constraint and be generated as `T`

**Comment by saibotma** [comment]
> An option is to generate false negatives only on records and POJOs (i.e. only nullable types are T?, the true positives, but non-nullable types, which are generated as T, would produce false negatives because they can still be null after an outer join, a union, etc.). But it would be a user choice, just like the !! operator. With this choice, TableField<R, T?> references would still always be nullable, only records and POJOs are affected, as they're mostly used with SELECT * projections of single tables (no joins)

is the solution we require.

**Comment by lukaseder** [comment]
Thanks for your comment @saibotma. This will definitely be included in jOOQ 3.15. While I have to repeat my disclaimers :), I definitely see how this can be *very* useful as a pragmatic 80/20 solution.

**Comment by vojtapol** [comment]
@lukaseder When you mention the false positive in case of left join I assume you mean something like:
```java
create.select()
  .from(BOOK)
  .leftJoin(AUTHOR).onKey()
  .where(BOOK.ID.eq(42))
  .fetchOneInto(AUTHOR)
```

Right? In case the book has no known author, the `AuthorRecord` would have all fields set to `null` which would lead to a false positive for fields that are marked as not null.

In my opinion though the call should not result in an `AuthorRecord` with false positives. Instead, the projection to `AUTHOR` should throw an exception. For a join like this, I would personally instead use the generic `Record` type and later project it into `AUTHOR` only in case where I know I am getting the full `AUTHOR` record:

```java
var record = create.select()
  .from(BOOK)
  .leftJoin(AUTHOR).onKey()
  .where(BOOK.ID.eq(42))
  .fetchOne();

if (record.get(AUTHOR.ID) != null) {
  var author = record.into(AUTHOR);
}
```

**Comment by lukaseder** [comment]
> When you mention the false positive in case of left join I assume you mean something like [...]

Yes.

> Instead, the projection to `AUTHOR` should throw an exception

No. (If only I had 0.02$ for every time I'm explaining these things 😁)

1. First, you'll get your exception, just a bit later. You won't gain that much by making it more explicit.
2. Should we really waste all these CPU cycles to reverse engineer how a random flat `Result` came to be? Results would have to keep a reference to the query that produced them, and we'd have to analyse it too.
3. And what if we don't have all the information about the producing query? E.g. when we use plain SQL templating, we might not be able to access nullability information. The behaviour would again be different, and thus be very confusing.
4. What makes you think it's so obvious to recognise "bad queries"? I invite you to do the exercise of going through the *entire jOOQ API* and assessing for each and every operator whether it can tamper with the nullability of an expression (hint: I think I can spend my time more wisely). Some examples, what about these queries?

    ```java
    create.select()
      .from(BOOK)
      .fullJoin(AUTHOR).onKey()
      .fetchInto(AUTHOR);
    ```

    Or this one?

    ```java
    create.select(val(null).as(AUTHOR.ID))
      .fetchOneInto(AUTHOR)
    ```

Who is jOOQ to judge anyone about their intentions? In SQL, I can have a structurally typed record that looks like a nominally typed record (e.g. that of a table). I can cast them back and forth (even in SQL, e.g. PostgreSQL PL/pgSQL and Oracle PL/SQL offer such tools), and pretend they're the same thing, irrespective of any constraints. At this point, it becomes once more clear that in SQL, nullability is not a type modifier but a constraint (which can even be deferred!). It works differently from what most of you wish it did, which is why this bike shed is being discussed so often time and again on all the channels.

The *only* way to get all of this right would be via an external DSL (i.e. native SQL), which allows for keeping track of nullability via the usual means a language spec and compiler can offer. In an internal DSL, this seems impossible, even with much more sophisticated languages than Java or Kotlin. (More on that below)

I've seen this fail many times in other libraries that eagerly and cleverly implemented some kotlin `T? vs T` or Scala `Option[T] vs T` magic only to compromise (a lot!) on SQL functionality (besides, even simple SQL queries become extremely hard to write in those APIs because of the cognitive friction this "type safety" introduces, let alone hard to compile). jOOQ does not make such a compromise. When it comes to this `null` bikeshed vs. completeness of vision in terms of SQL support, jOOQ will *always* choose SQL. 

Yes, 80% of developers write boring CRUD operations where super fancy SQL does not matter. But jOOQ has been made for the 20% as well. And even if you're part of those 80%, you might *occasionally* want a `NOT NULL DEFAULT CURRENT_TIMESTAMP` modifier on your column, and then what?

> In my opinion though the call should not result in an `AuthorRecord` with false positives

As always, when someone wants an improvement deep down in feature X, what they may have actually wanted is feature Y (https://xyproblem.info). You don't really want `AuthorRecord` to be ridden of false positives *in your specific examples*. What you want is an alternative jOOQ that cannot produce `AuthorRecord` with false positives *in the first place*.

Maybe we can offer this in the future, see https://github.com/jOOQ/jOOQ/issues/11049, https://github.com/jOOQ/jOOQ/issues/11054, but it will not work with the internal DSL. Perhaps, in the future, we may be able to offer an API that parses SQL and reflects on it (at compile time) to produce the correct nullability information for your records and data classes *per query*.

This issue here will not discuss that, and it will definitely ship with the offered compromise of occasionally getting these false positives. I mean, heck, just call `create.newRecord(AUTHOR)` and boom there you have it. False positives wherever you look!

**Comment by sgrimm** [comment]
> An option is to generate false negatives only on records and POJOs

Do POJOs and records need to solve this the same way? That seems to be kind of implied in some of the discussion but perhaps it's okay for POJOs to have a different approach to nullability than records, or at least the option of a different approach. For my use cases, I'd be happy if POJOs had non-nullable fields for `NOT NULL` columns and records had nullable ones.

**Comment by lukaseder** [comment]
@sgrimm It would be an option, yes, with the usual caveat of probably breaking the `<interfaces/>` feature: https://github.com/jOOQ/jOOQ/issues/10509

**Comment by pauliusrum** [comment]
We currently use java generator + annotations to achieve nullability in Kotlin. Kotlin compiler takes into consideration java annotations (Nonnull, nullable) and generates nullable or non nullable types accordingly.

I'm kinda confused why java generator + annotations are not interchangeable with kotlin generator.  I understand the left join reasoning but does it not apply to java generator + annotations the same way as it does to kotlin generator?

**Comment by lukaseder** [comment]
@pauliusrum Do you think those annotations should also be produced in kotlin code? Would that be idiomatic?

**Comment by pauliusrum** [comment]
> @pauliusrum Do you think those annotations should also be produced in kotlin code? Would that be idiomatic?

Adding annotations to kotlin code does not have the same effect as with java. The compiler ignores the annotations when they are added to kotlin code.

```kotlin
data class Foo(
    @Nonnull val bar: String?
)

fun main() {
    val foo = Foo(null) // even if the annotation is applied the compiler allows null value
    val bar: String = foo.bar // compilation error since the type is nullable
}
```

**Comment by lukaseder** [comment]
So, what are you trying to say, then?

The JavaGenerator behaviour defaults to generating `T!` types, which are the best default for jOOQ, IMO. But `T!` types are not denotable in kotlin, so the status quo is to default to `T?` types, which aren't wrong, but not user friendly.

Yes, this needs to be improved. But differently, not like the Java annotations...

**Comment by pauliusrum** [comment]
> Yes, this needs to be improved. But differently, not like the Java annotations...
Agree. I was just wondering why java annotations were introduced. What problems did they solve? The same ones that we are discussing in this issues or something else?

**Comment by lukaseder** [comment]
> Agree. I was just wondering why java annotations were introduced.

Well, this bikeshed can be painted in any language! 😉

**Comment by sgrimm** [comment]
> @sgrimm It would be an option, yes, with the usual caveat of probably breaking the `<interfaces/>` feature: #10509

Not sure if this totally addresses that concern, but nullable interface methods/properties can be implemented as non-nullable. For example:

```kotlin
interface NullableInterface {
  val property: String?
}

data class NonNullableImplementation(
  override val property: String
) : NullableInterface

val example = NonNullableImplementation("x")

// Property is of type String when accessed on implementation
assert(example.property.length == 1)

// Property is of type String? when accessed using interface
assert((example as NullableInterface).property?.length == 1)
```

So if I understand the problem correctly, the properties in the interface would be nullable, would be implemented as nullable in records as they are today, but would be non-nullable in POJOs.

**Comment by lukaseder** [comment]
@sgrimm Since `T <: T?`, I can see how covariant overloading a `val` covariantly can work. But we generate `var`, not `val` since it is desirable to be able to set values on interfaces and records, perhaps a bit less so on data classes, which may be immutable. There is no possibility of a covariant override of a `val` property, because of the setters. You can try it in your code. This doesn't work:

```kotlin
interface NullableInterface {
  var property: String?
}

data class NonNullableImplementation(
  override var property: String
) : NullableInterface
```

We could, of course, consider whether we generate `<immutablePojos/>` and `<immutableInterfaces/>`, but this does start to get really complicated, then...

**Comment by JohannesZick** [comment]
I was also thinking of this feature. I'm currently using the Kotlin generator, and I write a lot of `!!` after fetching data.

I see a big difference though between Records and POJOs (POKOs?). Records track what has been set, so if you try to read an unset field that should be non-null... what do you expect to happen? I don't think exceptions in these cases make anything easier to use. POJOs are different because they don't have this distinction. By making fields non-null, you loose the ability to construct incomplete POJOs, e.g. for inserts. But IMHO in that case you should use Records to mark which columns should be touched at all in the query.

Even in a POJO you might not have an id though. So, maybe columns that are marked as `GENERATE ALWAYS` or similar need to be nullable anyway? As much as I want this, it seems like a slippery slope into many many corner cases.

**Comment by lukaseder** [comment]
> As much as I want this, it seems like a slippery slope into many many corner cases.

My words :) I'm convinced most people in this discussion don't actually want what they think they want.

**Comment by JohannesZick** [comment]
OK, so what I think could work without too much confusion, but also limitations:

An option to make all POJO fields non-null using the same information as used for annotations. Not on Records, and on *all* fields for POJOs.

This limits usability of POJOs. You cannot easily use them for inserts or updates because generated values (e.g. ids) have to be set in the constructor. But you should really use Records there IMHO, exactly to make the distinction between "set to null" and "set to generated value".

The use case mentioned before with JOINs is I think even less relevant. These POJOs are meant to represent exactly one row in exactly one table. When you create a new table by joining, the POJO will not match and you should run your own.

**Comment by roguearos** [comment]
One potential solution here is to generate "complete" record classes which can delegate to "partial" record classes. This circumvents the [issue with overriding of nullable var](https://github.com/jOOQ/jOOQ/issues/10212#issuecomment-784188756).

```kotlin
class UsersRecordComplete(val record: UsersRecord) {
    var name 
        get() = record.name!!
        set(name: String) { record.name = name }
}
```

When queried through DSLContext we always get back UsersRecord instance and it is upto consumer to wrap them into "complete" records if they know for a fact that the row is complete. 

The DAOs can optionally wrap before returning based on a configuration option (which users can disable if they want to avoid the overhead of delegation).

**Comment by sgrimm** [comment]
An approach I've played with a bit in my application code, with some amount of success, is to use generics for context-sensitive nullability. Something along the lines of

```kotlin
// There is a nullable type parameter for each data type that occurs in a NOT NULL column.
class GenericUsersRecord<T1 : Long?, T2 : String?>(
    var id: T1,
    var name: T2,
    // This one can just be a regular String? since it maps to a nullable column.
    var optionalValue: String?
) {
  // This approach is unsafe if something keeps a reference to the same object with nullable types,
  // but is arguably fine for use in a call chain.
  fun asComplete(): CompleteUsersRecord {
    if (id == null || name == null) throw IllegalStateException("Nulls!")
    @Suppress("UNCHECKED_CAST") return this as CompleteUsersRecord
  }

  // This approach is safe but allocates a possibly-superfluous extra object.
  fun toComplete(): CompleteUsersRecord {
    return CompleteUsersRecord(
        id ?: throw IllegalStateException("id must not be null"),
        name ?: throw IllegalStateException("name must not be null"),
        optionalValue)
  }

  fun asNullable(): UsersRecord {
    @Suppress("UNCHECKED_CAST") return this as UsersRecord
  }

  fun toNullable(): UsersRecord {
    return UsersRecord(id, name, optionalValue)
  }
}

// Code that expects everything to be nullable continues to work as before.
typealias UsersRecord = GenericUsersRecord<Long?, String?>

// Code that expects complete records uses this type alias.
typealias CompleteUsersRecord = GenericUsersRecord<Long, String>
```

Of course, I am toying with this locally with the particular permutation of code generator options I happen to be using; maybe it breaks in some cases.

**Comment by lukaseder** [comment]
Related discussion when using the new jOOQ 3.15 `MULTISET` and ad-hoc data type conversion features: https://github.com/jOOQ/jOOQ/issues/12347

**Comment by strindberg** [comment]
We're very happy with our functional jOOQ mapping using Kotlin, and our repository code feels as elegant and minimal as it could possibly be. The only major remaining itch is the (lack of) null-safety of the generated code. Some type of "null-safe view" of non-null columns when using the new functional mapping possibilities in jOOQ would be of use for us.

Consider the following table:

```sql
CREATE TABLE event
(
    id         bigint NOT NULL GENERATED BY DEFAULT AS IDENTITY,
    message    text NOT NULL,
    PRIMARY KEY (id)
);
```

This maps to our data class:
```kotlin
data class Event(
    val id: Long,
    val message: String,
)
```

This is what our select looks like:
```kotlin
dsl.select(
   EVENT.ID,
   EVENT.MESSAGE,
).from(EVENT).where(EVENT.ID.eq(eventId)).fetchOne(mapping(::toEvent))

private fun toEvent(id: Long?, message: String?) = Event(id!!, message!!)
```

If there were some null-safe way of referring to what we select, we could get rid of the `toEvent` method and refer directly to the constructor in our select:

```kotlin
dsl.select(
   EVENT.ID,
   EVENT.MESSAGE,
).from(EVENT).where(EVENT.ID.eq(eventId)).fetchOne(mapping(::Event))
```

This pattern is of course found in a myriad places in all our repository code.

**Comment by lukaseder** [comment]
Thanks for your message, @strindberg. Can you create a new feature request for precisely that idea. I can experiment with an alternative set of `mapping` methods, written in kotlin, which would achieve that last idea, which isn't strictly related to this issue here.

**Comment by strindberg** [comment]
Done: #12934 . Let me know if there's any more details I should add.

**Comment by ukarlsson** [comment]
I use this method on the fields

```
fun <V, T: TableField<Record, V?>> T.asNotNull() = this as TableField<Record, V>
```
Example usage looks like

```
DSL.select(
   TABLE.FIELD.asNotNull()
).from(TABLE)
```

Definiton

**Comment by lukaseder** [comment]
https://github.com/jOOQ/jOOQ/issues/12934 has been implemented for jOOQ 3.18. It is still subject to change, but might do the trick for users who have written their own `data class` objects with non-null attributes, and wish to use them in mapping code, e.g. like this:

```kotlin
data class R (val i: Int, val s: Date?)
val x: List<R> = ctx.select(RECORD2.C1, RECORD2.C2).from(RECORD2).fetch(mappingIgnoreNulls(::R))
```

So, `mappingIgnoreNulls(...)` instead of `mapping(...)`. This is a quick win that eases some of the pain. I will think about further options related to this issue for jOOQ 3.18 as well, given the high number of upvotes on the issue.

**Comment by lukaseder** [comment]
While #12934 is a no-brainer (completely opt in utility for use with certain mapping calls), this issue here has been more controversial. The most controversial part is that some users would like a `TableField<R, T>` guarantee rather than `TableField<R, T?>` as it is today. I don't think I'll ever agree this is something jOOQ should do. A very simple case where these things break apart very quickly is functions or operators:

```kotlin

https://github.com/jOOQ/jOOQ/issues/13999

**Comment by lukaseder** [comment]
Small caveat when generating nullability: Defaulted (or identity) columns are nullable. There's no meaningful value users can provide when writing the value, see: https://github.com/jOOQ/jOOQ/issues/5128

**Comment by lukaseder** [comment]
A tricky discussion is the one about the primary constructor, which currently allows for creating an all-`null` record. This constructor is required by jOOQ's internals, where a record isn't populated *yet* before it is. But users shouldn't be able to call this constructor anymore. That might be done separately via https://github.com/jOOQ/jOOQ/issues/14002

**Comment by lukaseder** [comment]
From my perspective, this seems good to go for jOOQ 3.18. The feature can be activated using:

```kotlin
<kotlinNotNullPojoAttributes>true</kotlinNotNullPojoAttributes><kotlinNotNullRecordAttributes>true</kotlinNotNullRecordAttributes><kotlinNotNullInterfaceAttributes>true</kotlinNotNullInterfaceAttributes>
```

(Flags can be set individually, too). There's always risk of oversight of edge cases, so early adopter feedback would be greatly appreciated!

**Comment by robstoll** [comment]
@lukaseder looking forward to this, I see there is a 3.18 tag but I cannot find it on maven central yet. Is there already a plan to release it?

**Comment by lukaseder** [comment]
> I see there is a 3.18 tag but I cannot find it on maven central yet.

It's not there.

> Is there already a plan to release it?

Yes, we'll release it.

**Comment by lukaseder** [comment]
> I see there is a 3.18 tag

Oh, I see, you mean this?
https://github.com/jOOQ/jOOQ/releases/tag/version-3.18.0-sql2cypher

That's just a snapshot tag for this third party product:
https://github.com/neo4j-contrib/sql2cypher/issues/27

**Comment by mikehearn** [comment]
Hello,

Is it expected that this feature doesn't work for views? I've tried it out and it gets nullability right for table records but records generated from views are still considered to be fully nullable.

**Comment by lukaseder** [comment]
Thanks for your message @mikehearn. Can you please report a new issue with details? https://github.com/jOOQ/jOOQ/issues/new/choose

**Comment by vinothkumardhanasekar** [comment]
requesting for clarification , 
we were using javagenerator in our gradle file to generate the records , since our project is in kotlin , thought of moving to kotlingenerator, though I was using the isKotlinNotNullRecordAttributes = true in the build.gradle file, the jooq is generating attributes of nullable type , but the table column is non null, is this supported in 3.18.3 version ? jooq {
    version.set("3.18.3") } , please clarify thanks a lot.

**Comment by lukaseder** [comment]
I'm probably going to link to these issues almost daily by now. Anyone have an idea how this could be communicated very clearly and at the right spot? I take that generating Javadoc isn't the right approach here (it would be too noisy).

**Comment by mikehearn** [comment]
Hmm, perhaps you could generate a report file (html, markdown etc) as part of the build, which explains anything that the user may need to know about the generated code. It can start with explanations of nullability decisions, but later also include any other information that may not be immediately intuitive. The build plugins can print the path to the build report?

**Comment by lukaseder** [comment]
There's already log output. I don't think a separate file is needed. I have mixed feelings about these things. For some, it's useful. For others, it's just noise... And most people will probably come here to ask this question long before they consider turning on the log output (in whatever format).

**Comment by mikehearn** [comment]
I personally never look at logs unless I'm debugging something because I
tend to assume that they'll be full of debugging noise. The gradle logs are
useless for example. I think there's a difference between a human readable
report that's properly organized, and log files. For example the Gradle
build scans are done in an annoying way to drive Enterprise sales but the
basic concept of a browseable or renderable document that contains things
you probably should know, isn't a bad one.

**Comment by lukaseder** [comment]
Maybe. It's a lot of effort to make that document, though (in a way that it won't be perceived as noisy). And this particular bikeshed here isn't a great driver for starting the project. But certainly a thing to keep in mind should such a report ever be implemented.

**Comment by vinothkumardhanasekar** [comment]
Thanks for the comments, will enable these flags and build again & check. 

// Generate non-nullable types on POJO attributes, where column is not null. Default is false.
      IsKotlinNotNullPojoAttributes = true

      // Generate non-nullable types on Record attributes, where column is not null. Default is false.
      isKotlinNotNullRecordAttributes = true

      // Generate non-nullable types on interface attributes, where column is not null. Default is false.
      isKotlinNotNullInterfaceAttributes = true

**Comment by lukaseder** [comment]
@mikehearn Anyway, I like the general idea of a code generation report and will put it on the long term roadmap: https://github.com/jOOQ/jOOQ/issues/14981

**Comment by vinothkumardhanasekar** [comment]
sorry to bother again , additional question , will the following flags work with jooq gradle plugin ?  thanks

jooq {
    version.set("3.18.3")
    configurations {


// Generate non-nullable types on POJO attributes, where column is not null. Default is false.
IsKotlinNotNullPojoAttributes = true

  // Generate non-nullable types on Record attributes, where column is not null. Default is false.
  isKotlinNotNullRecordAttributes = true

  // Generate non-nullable types on interface attributes, where column is not null. Default is false.
  isKotlinNotNullInterfaceAttributes = true

**Comment by lukaseder** [comment]
@vinothkumardhanasekar All flags are supported by the third party gradle plugin (assumign you mean this one: https://github.com/etiennestuder/gradle-jooq-plugin). You may have to specify the plugin / XSD versions separately, see their docs:
https://github.com/etiennestuder/gradle-jooq-plugin#enforcing-the-jooq-configuration-xml-schema-version

**Comment by jacekgajek** [comment]
How to enable this with maven? I tried:
```
                <plugin>
                    <groupId>org.jooq</groupId>
                    <artifactId>jooq-codegen-maven</artifactId>
                    <version>${jooq.version}</version>
                    <executions>
                        <execution>
                            <id>jooq-codegen</id>
                            <phase>generate-sources</phase>
                            <goals>
                                <goal>generate</goal>
                            </goals>
                        </execution>
                    </executions>
                    <configuration>
                        <skip>${skipTests}</skip>
                        <jdbc>
                            <driver>org.postgresql.Driver</driver>
                            <url>jdbc:postgresql://${docker.hostName}:${jooq.port}/${jooq.database}</url>
                            <user>${jooq.user}</user>
                            <password>${jooq.password}</password>
                        </jdbc>
                        <generator>
                            <name>org.jooq.codegen.KotlinGenerator</name>
                            <isKotlinNotNullPojoAttributes>true</isKotlinNotNullPojoAttributes>
                            <isKotlinNotNullRecordAttributes>true</isKotlinNotNullRecordAttributes>
                            <isKotlinNotNullInterfaceAttributes>true</isKotlinNotNullInterfaceAttributes>


                            <database>
                                <name>org.jooq.meta.postgres.PostgresDatabase</name>
                                <includes>.*</includes>
                                <inputSchema>public</inputSchema>
                            </database>
                            <target>
                                <packageName>${mainPackage}.db_schema</packageName>
                                <directory>src/main/java-generated</directory>
                            </target>
                        </generator>
                    </configuration>
                </plugin>
```
but got error
```
 Unable to parse configuration of mojo org.jooq:jooq-codegen-maven:3.19.7:generate for parameter isKotlinNotNullPojoAttributes: Cannot find 'isKotlinNotNullPojoAttributes' in class org.jooq.meta.jaxb.Generator
```
The same with
```
                            <KotlinNotNullPojoAttributes>true</isKotlinNotNullPojoAttributes>
                            <KotlinNotNullRecordAttributes>true</isKotlinNotNullRecordAttributes>
                            <KotlinNotNullInterfaceAttributes>true</isKotlinNotNullInterfaceAttributes>
```
(without is). Tried lower-case too
I use version `3.19.7`

**Comment by mklnwt** [comment]
@lukaseder is there a way to select fields as non-nullable?

The comment in [KotlinGenerator](https://www.jooq.org/doc/3.20/manual/code-generation/codegen-output-languages/kotlingenerator/) was really helpful for understanding.
> While it is tempting to use non-nullability guarantees that are derived from your table definitions in generated code, it is important to remember that in SQL, there are numerous reasons why an expression that is based on a column declared as SQL NOT NULL can become NULL through the use of SQL operators, such as LEFT JOIN or UNION, etc. In particular, a column with a DEFAULT or IDENTITY expression is "nullable on write", so jOOQ will generate it as nullable despite a NOT NULL constraint!

Since it's encouraged to select only necessary fields, it would be great not having to do a non-null check for each field.

Simplified example, there might be more fields than just `name`.
```kotlin
dsl.select(BOOK.AUTHOR_ID, BOOK.NAME)
    .from(BOOK)
    .fetch { (authorId, name) -> id to name }
    .groupBy({ (authorId) -> authorId }, { (_, name) -> name })
```
or
```kotlin
dsl.select(BOOK.AUTHOR_ID, BOOK.NAME)
    .from(BOOK)
    .fetchGroups()
```
With the column `BOOK.NAME` being non-nullable, both of these give `Map<Long?, Collection<String?>>`.

My guess is that `dsl.select(BOOK.AUTHOR_ID.notNull)` to add a non-null check in the where clause won't be possible.
Enriching the nullability from a where clause also does not seem practical.
```kotlin
dsl.select(BOOK.AUTHOR_ID)
    .from(BOOK)
    .where(BOOK.AUTHOR_ID.isNotNull())
```

We're trying to avoid (code checks) `!!` in our code-base and adding `checkNotNull`s to every field is a lot of boilerplate. For `fetchGroups()` it's not even an option without additional mapping/filtering.

Would `BOOK.AUTHOR_ID.notNull` as an unchecked/unsafe syntactic sugar (with proper documentation) work here?

**Comment by lukaseder** [comment]
@mklnwt Please check out: https://github.com/jOOQ/jOOQ/issues/13999

> Enriching the nullability from a where clause also does not seem practical.

Well, it would be very interesting, but I believe I've only ever seen TypeScript or F# to offer a type system rich enough to do these kinds of magic tricks. It's certainly not possible in Java / Kotlin.

> Would `BOOK.AUTHOR_ID.notNull` as an unchecked/unsafe syntactic sugar (with proper documentation) work here?

You can add such an extension method yourself, very easily. It's just another way of writing `!!`, of course. Alternatively, as I remind folks ever so often, all this pain only happens because people use the `KotlinGenerator`, which can't generate platform types (the only honest type as far as jOOQ goes). You wouldn't have any of this with the `JavaGenerator`.

**Comment by mklnwt** [comment]
@lukaseder can you elaborate what you mean by "platform types"?

As long as the `JavaGenerator` doesn't add `@NotNull` (which I assume it doesn't based on the fact the a field might always be nullable), this wouldn't solve the nullability issue.

**Comment by lukaseder** [comment]
> [@lukaseder](https://github.com/lukaseder) can you elaborate what you mean by "platform types"?

This is a core concept of the kotlin language:

- https://kotlinlang.org/spec/type-system.html#platform-types

Here's a feature request to add denotability of platform types to the language:

- https://youtrack.jetbrains.com/issue/KT-56899

> As long as the `JavaGenerator` doesn't add `@NotNull` (which I assume it doesn't based on the fact the a field might always be nullable),

It does the right thing.

> this wouldn't solve the nullability issue.

I'm afraid, I don't know what you mean by this.

**Comment by mklnwt** [comment]
Basically the issue is not really the `KotlinGenerator`. It's that we can't tell if a column could be null or not (regardless of the language used).

It's generally somewhat simple to get around this by using `!!` or `checkNotNull`.

I also like your suggestion of adding 2 fields (one for nullable, one for non-nullable).
But I would keep it nullable by default as it is right now and instead add an additional non-nullable one.
Or as I suggested adding a `.notNull` extension that can convert any field into a non-nullable one. Then there wouldn't be the need to generate duplicate fields for each column.

Currently I don't see an option to convert fields this way ourselves. Plus it would be helpful to let the parser check fields marked as non-nullable.

**Comment by lukaseder** [comment]
> Basically the issue is not really the `KotlinGenerator`. It's that we can't tell if a column could be null or not (regardless of the language used).

It is very much an issue of the `KotlinGenerator`. People who use the `JavaGenerator` (in Kotlin or in Java) tend to simply not worry about this topic very much or even at all (including Kotlin developers). This is only an issue for people who use the `KotlinGenerator` which produces nullable types all over the place, because it cannot denote the fact that we simply "don't know" or "don't care" (platform type).

> I also like your suggestion of adding 2 fields (one for nullable, one for non-nullable).

I don't think that's what's going to be done here

> Or as I suggested adding a `.notNull` extension that can convert any field into a non-nullable one. Then there wouldn't be the need to generate duplicate fields for each column.

But then again, anyone can add any extension function that they like, so I'm not 100% convinced that this is something jOOQ should offer out of the box.

> Currently I don't see an option to convert fields this way ourselves.

I don't understand. Why not just write this very simple extension function? Why do you rely on this being offered out of the box?

```kotlin
@Suppress("UNCHECKED_CAST")
fun <T> Field<T?>.nonNull(): Field<T> = this as Field<T>
```

This works for me, then:

```kotlin
val r: Record1<Int> =
ctx().select(TEST.ID.nonNull())
    .from(TEST)
    .fetchSingle()
```

**Comment by mklnwt** [comment]
Sorry for the delayed response. I finally had some time to test this a bit more.

While this generally works and satisfies the compiler, it leads to ugly `NullPointerException`s.

Maybe it would make sense to move this into a separate issue.
My idea/request would be to add it as a feature to JOOQ to do the check and raise an appropriate exception with message if failed.

If this
```kotlin
val r: Record1<Int> =
ctx().select(TEST.ID.nonNull())
    .from(TEST)
    .fetchSingle()
```
could behave something like
```kotlin
val r: Record1<Int> =
ctx().select(TEST.ID)
    .from(TEST)
    .fetchSingle { (id) ->
        if (id == null) throw UnexpectedNullException(TEST.ID)
        id
    }
```
it would be ideal.

Basically let `.nonNull()` not only cast the field, but also register a null check mapper.

Thinking about it, my suggestion would not only help Kotlin. It would give a declarative way for Java to expect values to be present.

**Comment by lukaseder** [comment]
Nullability is an endless bikeshed. 

Luckily, Kotlin users can easily add an extension function. Java (and Kotlin) users can also add a converter.

I don't think this belongs in jOOQ's core, as I really don't see the value of this `UnexpectedNullException` over just `NullPointerException`, which is just as straightforward.

**Comment by mklnwt** [comment]
> Nullability is an endless bikeshed.

In terms of type safety it might be, but proper validation shouldn't be.

> Luckily, Kotlin users can easily add an extension function. Java (and Kotlin) users can also add a converter.

Thanks for this suggestion, I think this would do the trick

```kotlin
fun <T> Field<T?>.notNull(): Field<T> = this.convertFrom {
    checkNotNull(it) { "Field [$qualifiedName] expected to not be null" }
}
```

> I don't think this belongs in jOOQ's core, as I really don't see the value of this UnexpectedNullException over just NullPointerException, which is just as straightforward.

I do agree that we might not need a separate exception. However it would be helpful for developers to get a better/earlier exception message.

```kotlin
val r: Record1<Int> =
ctx().select(TEST.FOO.nonNull(), TEST.BAR.nonNull(), TEST.BAZ.nonNull())
    .from(TEST)
    .fetchSingle { (foo, bar, baz) -> TestResult(foo, bar, baz) }
```
could result in the `NullPointerException` being raised anywhere in your application on usage.

Of course arguably the author of code above could/should probably add null checks in the mapping to `TestResult`.
`notNull()` would just be a very straight-forward way of achieving such. Imo this would be a very common check.

Anyway, for now I can just add above solution as a library to our codebase.

**Comment by lukaseder** [comment]
> Of course arguably the author of code above could/should probably add null checks in the mapping to `TestResult`.

If you're using Kotlin, the Kotlin compiler will add the null check to your constructor call:

```kotlin
class X(a: String?, b: String) 

val a: String? = null;
val b: String? = null;

fun main() {
    X(
        a,
        b!!
    )
}
```

Resulting in:

```
Exception in thread "main" java.lang.NullPointerException
 at FileKt.main (File.kt:9) 
 at FileKt.main (File.kt:-1) 
 at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (:-2)
```

You don't have to do anything
